Support grizzly-memcached's monitoring features using Grizzly JMX

carryel opened the pull request:
JMX Monitoring functions have been added by applying the Grizzly JMX structure.

The following objects are monitored:
- GrizzlyMemcachedCache
- BaseObjectPool
- KeyedObject (which is generally related to network connection management in the basic object pool).

Most statistics are provided as JMX CompositeData.

PS) Since several patches were previously included in a single PR(https://github.com/eclipse-ee4j/glassfish-grizzly-memcached/pull/14), I submit this by extracting only the additional features again.

carryel committed:
+ Implemented "Support grizzly-memcached's monitoring features using Grizzly JMX"(https://github.com/eclipse-ee4j/grizzly-memcached/issues/13)
+ (Trivial) Fixed several log messages and line alignment.
+ (Trivial) Updated pom's library version.

carryel committed:
Implemented "Support grizzly-memcached's monitoring features using Grizzly JMX"(https://github.com/eclipse-ee4j/grizzly-memcached/issues/13)
+ removed unnecessary comments(local test codes).

carryel committed:
Implemented "Support grizzly-memcached's monitoring features using Grizzly JMX"(https://github.com/eclipse-ee4j/grizzly-memcached/issues/13)
+ added several ManagedOperation

carryel committed:
Implemented "Support grizzly-memcached's monitoring features using Grizzly JMX"(https://github.com/eclipse-ee4j/grizzly-memcached/issues/13)
+ fixed testcase's failure.

carryel committed:
Implemented "Support grizzly-memcached's monitoring features using Grizzly JMX"(https://github.com/eclipse-ee4j/grizzly-memcached/issues/13)
+ For errors that occur when enabling/disabling JMX, only logs are left and the errors are ignored.

carryel committed:
Implemented "Support grizzly-memcached's monitoring features using Grizzly JMX"(https://github.com/eclipse-ee4j/grizzly-memcached/issues/13)
+ Added stats using JMX's CompositeData.

carryel committed:
Implemented "Support grizzly-memcached's monitoring features using Grizzly JMX"(https://github.com/eclipse-ee4j/grizzly-memcached/issues/13)
+ Added stats(Keyed object which is generally related to network connection managed in the basic object pool).

carryel committed:
- Updated the major version of Grizzly framework (2.4.4 -> 4.0.2).
- Changed the minimum java version to 11 or higher.
- Replaced the use of NullaryFunction, which has been removed since Grizzly framework 4.0.1, with Supplier (related issue https://github.com/eclipse-ee4j/grizzly/pull/2179).
- Updated the minor version of grizzly-memcached (1.3.x -> 1.4.x).

carryel committed:
Implemented "Support grizzly-memcached's monitoring features using Grizzly JMX"(https://github.com/eclipse-ee4j/grizzly-memcached/issues/13)
+ Fixed an issue where the previous pool was monitored instead of the newly created pool when the pool was dynamically destroyed and recreated.
+ Remove unused(duplicated) mbean attributes.

carryel committed:
Implemented "Support grizzly-memcached's monitoring features using Grizzly JMX"(https://github.com/eclipse-ee4j/grizzly-memcached/issues/13)
+ (Trivial) For convenience when connecting to monitor, the connection pool name is simply changed to the client name.

carryel committed:
Implemented "Support grizzly-memcached's monitoring features using Grizzly JMX"(https://github.com/eclipse-ee4j/grizzly-memcached/issues/13)
+ Changed jdk to 17 and upgraded the versions of other libraries.

carryel committed:
Implemented "Support grizzly-memcached's monitoring features using Grizzly JMX"(https://github.com/eclipse-ee4j/grizzly-memcached/issues/13).

pzygielo commented on src/main/java/org/glassfish/grizzly/memcached/jmx/GrizzlyMemcachedCache.java:
```diff
@@ -0,0 +1,245 @@
+    }
+
+    @ManagedData(name="Connection Stat")
+    private static class CompositeConnectionStat {
```
This one is `private`, but `o.g.g.m.p.jmx.KeyedObject.CompositeObjectStat` and  `o.g.g.m.p.jmx.BaseObjectPool.CompositeObjectPoolStat` are `public`. If there is no specific reason - for consistency, could it be synchronized?

carryel replied:
Since it has no effect whether private class or public class, I synchronized it to private one.

pzygielo commented on src/main/java/org/glassfish/grizzly/memcached/jmx/GrizzlyMemcachedCache.java:
```diff
@@ -0,0 +1,245 @@
+    @ManagedOperation(id = "grizzly-memcached-cache-connection-size", impact = Impact.INFO)
+    @Description("The total number of connections currently idle and active in this pool or a negative value if unsupported.")
+    @ParameterNames({"hostname", "port"})
+    public int getConnectionSize( final String hostname, final int port) {
```
If this file is edited - let's remove ` ` before first `final`.

carryel replied:
fixed.

carryel committed:
Implemented "Support grizzly-memcached's monitoring features using Grizzly JMX"(https://github.com/eclipse-ee4j/grizzly-memcached/issues/13).

Applied the following review comments:
+ https://github.com/eclipse-ee4j/glassfish-grizzly-memcached/pull/20#discussion_r2346894670
+ https://github.com/eclipse-ee4j/glassfish-grizzly-memcached/pull/20#discussion_r2346895630
+ https://github.com/eclipse-ee4j/glassfish-grizzly-memcached/pull/20#discussion_r2346896127

pzygielo commented on src/main/java/org/glassfish/grizzly/memcached/pool/jmx/BaseObjectPool.java:
```diff
@@ -0,0 +1,156 @@
+    @Override
+    protected void onRegister(GrizzlyJmxManager mom, GmbalMBean bean) {
+        this.mom = mom;
+        rebuildSubTree();
```
Could `mom` be passed to `rebuildSubTree` and `this.mom` field removed?

carryel replied:
I think it would be better to modify it to the flow you suggested.

Actually, this pattern ended up like this because I was porting Grizzly's monitoring logic. 😄

https://github.com/eclipse-ee4j/glassfish-grizzly/blob/7ea0c5745bd0d65d4a939fa33a7930faf8603cf4/modules/monitoring/http-server/src/main/java/org/glassfish/grizzly/http/server/jmx/HttpServer.java#L74
https://github.com/eclipse-ee4j/glassfish-grizzly/blob/7ea0c5745bd0d65d4a939fa33a7930faf8603cf4/modules/monitoring/http-server/src/main/java/org/glassfish/grizzly/http/server/jmx/NetworkListener.java#L85
https://github.com/eclipse-ee4j/glassfish-grizzly/blob/7ea0c5745bd0d65d4a939fa33a7930faf8603cf4/modules/monitoring/grizzly/src/main/java/org/glassfish/grizzly/nio/transport/jmx/NIOTransport.java#L99

I think HttpServer and NetworkListener could also be improved in the future.

NIOTransport likely needed the field because of the call below: 
https://github.com/eclipse-ee4j/glassfish-grizzly/blob/7ea0c5745bd0d65d4a939fa33a7930faf8603cf4/modules/monitoring/grizzly/src/main/java/org/glassfish/grizzly/nio/transport/jmx/NIOTransport.java#L328

PS) This is a lot of information, but I've included an explanation for future reference. :-)

carryel committed:
Implemented "Support grizzly-memcached's monitoring features using Grizzly JMX"(https://github.com/eclipse-ee4j/grizzly-memcached/issues/13).

Applied the following review comments:
+ https://github.com/eclipse-ee4j/glassfish-grizzly-memcached/pull/20#discussion_r2347268576
+ https://github.com/eclipse-ee4j/glassfish-grizzly-memcached/pull/20#discussion_r2347270166
